common/scala/src/main/scala/org/apache/openwhisk/core/entity/ParameterEncryption.scala

```diff
@@ -0,0 +1,147 @@
+    case 32 => new Aes256(storageConfig.getKeyBytes())
+    case 0  => new NoopCrypt
+    case _ =>
+      throw new IllegalArgumentException(
+        s"Only 0, 16 and 32 characters support for key size but instead got ${storageConfig.getKeyBytes().length}")
+  }
+  def lock(params: Parameters): Parameters = {
+    new Parameters(
+      params.getMap
+        .map(({
+          case (paramName, paramValue) if paramValue.encryption == JsNull =>
+            paramName -> enc.encrypt(paramValue)
+          case (paramName, paramValue) => paramName -> paramValue
+        })))
+  }
+  def unlock(params: Parameters): Parameters = {
+    new Parameters(
+      params.getMap
+        .map(({
+          case (paramName, paramValue)
+              if paramValue.encryption != JsNull && paramValue.encryption.convertTo[String] == enc.name =>
+            paramName -> enc.decrypt(paramValue)
+          case (paramName, paramValue) => paramName -> paramValue
+        })))
+  }
+}
+
+private trait AesEncryption extends encrypter {
+  val key: Array[Byte]
+  val ivLen: Int
+  val name: String
+  private val tLen = key.length * 8
+  private val secretKey = new SecretKeySpec(key, "AES")
+
+  private val secureRandom = new SecureRandom()
+
+  def encrypt(value: ParameterValue): ParameterValue = {
+
+    val iv = new Array[Byte](ivLen)
+    secureRandom.nextBytes(iv)
+    val gcmSpec = new GCMParameterSpec(tLen, iv)
+    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
+    cipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmSpec)
+    val clearText = value.value.convertTo[String].getBytes(StandardCharsets.UTF_8)
+    val cipherText = cipher.doFinal(clearText)
+
+    val byteBuffer = ByteBuffer.allocate(4 + iv.length + cipherText.length)
+    byteBuffer.putInt(iv.length)
+    byteBuffer.put(iv)
+    byteBuffer.put(cipherText)
+    val cipherMessage = byteBuffer.array
+    ParameterValue(JsString(Base64.getEncoder.encodeToString(cipherMessage)), value.init, JsString(name))
+  }
+
+  def decrypt(value: ParameterValue): ParameterValue = {
+    val cipherMessage = value.value.convertTo[String].getBytes(StandardCharsets.UTF_8)
+    val byteBuffer = ByteBuffer.wrap(Base64.getDecoder.decode(cipherMessage))
+    val ivLength = byteBuffer.getInt
+    if (ivLength != ivLen) {
+      throw new IllegalArgumentException("invalid iv length")
+    }
+    val iv = new Array[Byte](ivLength)
+    byteBuffer.get(iv)
+    val cipherText = new Array[Byte](byteBuffer.remaining)
+    byteBuffer.get(cipherText)
+
+    val gcmSpec = new GCMParameterSpec(tLen, iv)
+    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
+    cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec)
+    val plainTextBytes = cipher.doFinal(cipherText)
+    val plainText = new String(plainTextBytes, StandardCharsets.UTF_8)
+    ParameterValue(JsString(plainText), value.init)
+  }
+
+}
+
+private case class Aes128(val key: Array[Byte], val ivLen: Int = 12, val name: String = "aes128")
```

Comment:
are there string constants to use for the values from the cipher libraries - if so should the code use those?

Reply:
This is just the name that will be stored alongside the parameters, they are only used by OW and are meant to be unique. The reason for this not being tied to some java constant is that there may have to be, due to incorrect handling on my part, aes128-1.

Reply:
Also this gives someone a place, if they wanted to register a new encrypter called, "aes-256-namespace" for instance. Which then would allow someone to add keys to a namespace and resolve it there. It could fall back to the aes256 if there was no key configured.